### PR TITLE
Linear Probing

### PR DESCRIPTION
**Pull request by Negiiiin**
# PR Type
[Feature]

# Short Description
Linear probing task

### CODE REVIEW

File: mmlearn/tasks/__init__.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 from mmlearn.tasks.contrastive_pretraining import ContrastivePretraining
 from mmlearn.tasks.ijepa import IJEPA
+from mmlearn.tasks.linear_probing import LinearClassifierModule
```
Comment (fcogidi) [inline mmlearn/tasks/__init__.py:5]:
```suggestion
from mmlearn.tasks.linear_probing import LinearClassifier
```

File: mmlearn/tasks/__init__.py [hunk 2]
```diff
@@ -11,4 +12,5 @@
     "IJEPA",
     "ZeroShotCrossModalRetrieval",
     "ZeroShotClassification",
+    "LinearClassifierModule",
```
Comment (fcogidi) [inline mmlearn/tasks/__init__.py:15]:
```suggestion
    "LinearClassifier",
```

File: mmlearn/tasks/linear_probing.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+            if state_dict:
+                model["rgb"].load_state_dict(state_dict, strict=True)
+                print("Encoder state dict loaded successfully")
+        except Exception as e:
+            print(f"Error loading state dict: {e}")
+    return model["rgb"]
+
+
+@store(group="task", provider="mmlearn")
+class LinearClassifierModule(L.LightningModule):
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:89]:
```suggestion
class LinearClassifier(L.LightningModule):
```

File: mmlearn/tasks/linear_probing.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+    References
+    ----------
+    [1] He, K., Chen, X., Xie, S., Li, Y., Doll'ar, P., & Girshick, R.B. (2021).
+        Masked Autoencoders Are Scalable Vision Learners. 2022 IEEE/CVF Conference
+        on Computer Vision and Pattern Recognition (CVPR), 15979-15988.
+    """
+
+    def __init__(
+        self,
+        # encoder: torch.nn.Module,
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:157]: Please remove this line.

File: mmlearn/tasks/linear_probing.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        encoder: nn.Module,
+        model_checkpoint_path: Optional[str],  # change name
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:159]:
```suggestion
        model_checkpoint_path: Optional[str],
```

File: mmlearn/tasks/linear_probing.py (hunk 3 above)
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:89]: Inherit from [TrainingTask](https://github.com/VectorInstitute/mmlearn/blob/main/mmlearn/tasks/base.py#L15) instead.

File: mmlearn/tasks/linear_probing.py [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+        super().__init__()
+        assert task in ["binary", "multiclass", "multilabel"], (
+            f"Invalid task type: {task}. "
+            "Expected one of ['binary', 'multiclass', 'multilabel']."
+        )
+
+        self.modality = modality
+
+        self.encoder: nn.Module = extract_vision_encoder(
+            encoder, model_checkpoint_path, keys_to_rename={"encoders.rgb.": ""}
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:185]: Expose `keys_to_remove`, `keys_to_rename` and `keys_to_ignore` during initialization of the `LinearClassifier` class and pass those parameters directly to this function.

File: mmlearn/tasks/linear_probing.py [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+        )
+
+        linear_layer = MLP(num_output_features, num_classes, hidden_dims)
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:188]: Expose the rest of the parameters for constructing an MLP (e.g. adding activation functions and drop out layers when using multiple linear layers. This may negate the need for the `pre_classifier_batch_norm` argument.

File: mmlearn/tasks/linear_probing.py [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+        self.classifier = linear_layer
+
+        self.freeze_encoder = freeze_encoder
+        self.num_classes = num_classes
+
+        if self.freeze_encoder:
+            for param in self.encoder.parameters():
+                param.requires_grad = False
+
+        self.loss_fn = nn.CrossEntropyLoss()
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:204]: Use `nn.BCEWithLogitsLoss` for multi-label tasks.

File: mmlearn/tasks/linear_probing.py [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+                "precision": Precision(
+                    task=task, num_labels=num_classes, average="macro"
+                ),
+                "recall": Recall(task=task, num_labels=num_classes, average="macro"),
+                "f1_score": F1Score(task=task, num_labels=num_classes, average="macro"),
+                "auc": AUROC(task=task, num_labels=num_classes),  # AUC for multilabel
+            }
+
+        else:  # binary
+            # Accuracy and other metrics for binary classification
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:246]: For creating metrics, try to make [this](https://github.com/VectorInstitute/mmlearn/blob/main/mmlearn/tasks/zero_shot_classification.py#L248) a standalone function that can be reused by the `LinearClassifier` and `ZeroShotClassification` classes. Currently, you can directly reuse it by calling `ZeroShotClassification._create_metrics...`.

File: mmlearn/tasks/linear_probing.py [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+        """
+        with torch.no_grad() if self.freeze_encoder else nullcontext():
+            x = self.encoder(x)
+        return self.classifier(x[0])
+
+    def _get_logits_and_labels(
+        self, batch: Dict[str, torch.Tensor]
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Return the logits and labels for a batch of data."""
+        x: torch.Tensor = batch
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:287]: The input data corresponding to `modality` is not being extracted from the batch.

File: mmlearn/tasks/linear_probing.py [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+        loss = self._compute_loss(batch)
+
+        if loss is None:
+            raise ValueError("The loss function must be provided for training.")
+
+        self.log("train/loss", loss, prog_bar=True)
+
+        return loss
+
+    def validation_step(
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:331]: For completeness, please also add a `test_step`.

File: mmlearn/tasks/linear_probing.py [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+        mapping
+            old prefixes to new replacements (default removes 'encoders.rgb.').
+        keys_to_ignore (Optional[list]): List of keys to ignore when loading the
+        state_dict.
+
+    Returns
+    -------
+        nn.Module: The vision encoder module extracted and initialized.
+    """
+    model: L.LightningModule = hydra.utils.instantiate(encoder)
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:46]: The encoder should already be instantiated when passed to the method, no?

File: mmlearn/tasks/linear_probing.py [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+                if k.startswith(old_prefix)
+            }
+
+        try:
+            if state_dict:
+                model["rgb"].load_state_dict(state_dict, strict=True)
+                print("Encoder state dict loaded successfully")
+        except Exception as e:
+            print(f"Error loading state dict: {e}")
+    return model["rgb"]
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:85]: Please generalize this as well. That is, allow the user to specify which key, if any, to extract.

File: mmlearn/tasks/linear_probing.py [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+        # Remove unwanted keys
+        if keys_to_remove:
+            state_dict = {
+                k: v for k, v in state_dict.items() if k not in keys_to_remove
+            }
+
+        # Ignore specific keys
+        if keys_to_ignore:
+            state_dict = {
+                k: v for k, v in state_dict.items() if k not in keys_to_ignore
```
Comment (fcogidi) [inline mmlearn/tasks/linear_probing.py:67]: This is removing the keys from the `state_dict`

File: projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml
Comment (fcogidi) [inline projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml:1]: Is the intention here to train a single linear classifier on all the datasets at once or a single linear classifier for each dataset?

File: projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+  test:
+    batch_size: 64
+    num_workers: 4
+    shuffle: False
+
+task:
+  task: multiclass
+  num_classes: 4
+  num_output_features: 512
+  hidden_dims: [256]
```
Comment (fcogidi) [inline projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml:269]: Why add hidden layers? Are you following a certain recipe?

File: projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+  modality: rgb
+  encoder_checkpoint_path: /path/to/checkpoint
```
Comment (fcogidi) [inline projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml:271]: You can make this a required parameter like so:

```suggestion
  encoder_checkpoint_path: ???
```

File: projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml
Comment (Negiiiin) [reply projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml:1]: single linear classifier for each dataset. This will work with one dataset each time

Comment (fcogidi) [inline projects/med_benchmarking/configs/experiment/linear_probing_eval.yaml:1]: @Negiiiin Please update this config file to match the current implementation of the `LinearEvaluation` task.
